man/man8/osindications.8
.\" Copyright (c) 2019 Johnothan King. All rights reserved.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in all
.\" copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.\" SOFTWARE.
.\"
.Dd May 20, 2019
.Dt OSINDICATIONS 8
.Os
.Sh NAME
.Nm osindications
.Nd Set OsIndications in
.Em /sys/firmware/efi/efivars
for booting into firmware setup
.Sh SYNOPSIS
.Nm osindications
.op Fl u?
.Sh DESCRIPTION
This program controls whether or not
.Nm OsIndications
in efivars is set to boot into
the system firmware's on the system's next boot.
When run without any flags or options,
.Nm osindications
will set
.Nm OsIndications
to boot into the system firmware on the next boot.

This program accepts the following flags:

.Nm -u, --unset
Unset the 0x0000000000000001 UEFI variable in
.Nm OsIndications
to reverse any previous actions that were taken.

.Nm -?, --help
Show
.Nm osindications
usage information.
.Sh FILES
.Em /sys/firmware/efi/efivars
This is the efivarfs that contains the system's UEFI variables.
Non-standard variables are made immutable by default.

.Em /sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c
The
.Nm OsIndicators
UEFI variable that
.Nm osindicators
writes to.
If this file is immutable,
.Nm osindicators
will fail to set the OsIndications UEFI variable.
This can be fixed by using
.Nm chattr(1).

.Em /sys/firmware/efi/efivars/OsIndicationsSupport-8be4df61-93ca-11d2-aa0d-00e098032b8c
This is the UEFI variable that tells
.Nm osindications
if the system supports rebooting into the firmware's interface from the current OS.
.Sh NOTES
This program has not yet been ported to
.Nm FreeBSD.
As of the time of writing,
.Nm osindicators
only runs on
.Nm Linux.
.Sh SEE ALSO
efivar(1)
.Sh AUTHOR
Johnothan King


cmd/osindications.c
/*
 * Copyright (c) 2019 Johnothan King. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/*
 * osindications - This program sets OsIndications for booting into firmware setup.
 */

#include "inc.h"

int main(int argc, char *argv[]) {

	// Error checks
	if(getuid() != 0) {
		printf(RED "* Permission denied!" RESET "\n");
		return 1;
	} else if(access("/sys/firmware/efi/efivars/OsIndicationsSupported-8be4df61-93ca-11d2-aa0d-00e098032b8c", R_OK) != 0) {
		printf(RED "* This system does not support OsIndications!" RESET "\n");
		return 1;
	}

	// Long options and first_bytes variable
	size_t first_bytes = 4;
	struct option long_options[] = {
		{ "unset", no_argument, 0, 'u' },
		{ "help",  no_argument, 0, '?' },
	};

	// Parse options
	int args;
	while((args = getopt_long(argc, argv, "u?", long_options, NULL)) != -1) {
		switch(args) {

			// Display usage info
			case '?':
				printf("Usage: %s [-u?]\n", __progname);
				printf("  -u, --unset            Revert changes made by osindications\n");
				printf("  -?, --help             Show this usage information\n");
				return 1;

			// Unset OsIndications
			case 'u':
				first_bytes++;
				break;
		}
	}

	// Write efi_attr (0x0000000000000007)
	FILE *fd = fopen("/sys/firmware/efi/efivars/OsIndications-8be4df61-93ca-11d2-aa0d-00e098032b8c", "w+");
	unsigned long efi_attr = 0x0000000000000007;
	if(fwrite(&efi_attr, 1, first_bytes, fd) == 0) {
		fclose(fd);
		printf(RED "* Failed to write the changes to OsIndications!" RESET "\n");
		return 1;
	}

	// Write efi_boot (0x0000000000000001)
	if(first_bytes == 4) {
		unsigned long efi_boot = 0x0000000000000001;
		fwrite(&efi_boot, 1, 1, fd);
	}

	// Close the file and notify the user of the change
	fclose(fd);
	sync();
	if(first_bytes == 4) {
		printf(CYAN "* " WHITE "This system will now boot into the firmware's UI the next time it boots." RESET "\n");
		printf(CYAN "* " WHITE "Run `osindications --unset` to revert this change." RESET "\n");
	} else {
		printf(CYAN "* " WHITE "This system will not boot into the firmware's UI the next time it boots." RESET "\n");
		printf(CYAN "* " WHITE "Any prior changes made by osindications have been reverted." RESET "\n");
	}

	return 0;
}
